mongoengine/queryset/base.py

```diff
@@ -1833,26 +1833,6 @@ def lookup(obj, name):
 
         return tuple(data)
 
-    def _get_as_pymongo(self, doc):
-        """Clean up a PyMongo doc, removing fields that were only fetched
-        for the sake of MongoEngine's implementation, and return it.
-        """
-        # Always remove _cls as a MongoEngine's implementation detail.
-        if '_cls' in doc:
-            del doc['_cls']
-
-        # If the _id was not included in a .only or was excluded in a .exclude,
-        # remove it from the doc (we always fetch it so that we can properly
-        # construct documents).
-        fields = self._loaded_fields
-        if fields and '_id' in doc and (
-            (fields.value == QueryFieldList.ONLY and '_id' not in fields.fields) or
-            (fields.value == QueryFieldList.EXCLUDE and '_id' in fields.fields)
```

Comment:
Regarding this old piece of code that was removed, I think it makes sense to keep part of it so that:
- using .only('field1').as_pymongo() still returns the "id" (so that it is consistent with mongoengine .only)
- using .exclude('id', 'field1').as_pymongo() does not return the "id"

Make sure those 2 cases are tested

Reply:
That was exactly the idea of removing all of the logic. It was removing the "id" field if you didn't explicitly ask for it (which is not how it works in pymongo).
Both cases are now tested in `test_as_pymongo_json_limit_fields` and I also added now another explicit check for your first bullet